libdnf/dnf-context.c

```diff
@@ -1655,6 +1688,9 @@ dnf_context_run(DnfContext *context, GCancellable *cancellable, GError **error)
     if (!dnf_state_done(priv->state, error))
         return FALSE;
 
+    if (priv->downloadonly)
+        goto out;
```

Comment:
I think you probably need to change state steps calculation for the downloadonly case, as it's going to have less steps to do in the early exist case.